### DISCUSSION TITLE
Issues with subtypes of types

### DISCUSSION THREAD

**Discussion by povilasb** [microsoft/pyright#3390] [General]
I'm having hard time to grasp if this is an issue with pyright or I'm misinterpreting Python type system.
Consider this small example:

```py
from typing import TypeVar, Sequence, Type

_T = TypeVar("_T", bound=Type[Exception])
_MaybeSequence = _T | Sequence[_T]


class HttpError(Exception):
    pass


def retry(errs: _MaybeSequence[type[Exception]]):
    pass


# both of these cases fail the type checks
retry(HttpError)
retry(Exception)
```

pyright (`1.1.240`) errors with

```sh
  /tmp/main.py:15:7 - error: Argument of type "Type[HttpError]" cannot be assigned to parameter "errs" of type "_MaybeSequence[Type[Exception]]" in function "retry"
    Type "Type[HttpError]" cannot be assigned to type "_MaybeSequence[Type[Exception]]"
      "Type[type]" is incompatible with "Type[Exception]"
      "Type[type]" is incompatible with "Type[Sequence[Type[Exception]]]" (reportGeneralTypeIssues)
```

"Type[type]" sounds suspicious.

Maybe subtypes of types are not even supposed to work this way?

I can check them during the runtime though
```py
print(issubclass(HttpError, Exception))
True
```
Also, `pyright` in this case says
```py
error: Unnecessary issubclass call; "Type[HttpError]" is always a subclass of "Exception" (reportUnnecessaryIsInstance)
```

**Comment by erictraut** [comment]
Here's the proper way to type this:

```python
from typing import TypeVar, Sequence, Type

_T = TypeVar("_T", bound=Exception)
_MaybeSequence = Type[_T] | Sequence[Type[_T]]

class HttpError(Exception):
    pass

def retry(errs: _MaybeSequence[Exception]):
    pass

retry(HttpError)
retry(Exception)
```

This approach works with both mypy and pyright, whereas the original example fails with both with similar errors.

**Reply by povilasb** [reply]
I see, thank you, Eric.

Any hints on why it doesn't work with `type[T]`?

Cause now I can't reuse a generic `_MaybeSequence` with types and rather have to maintain two flavours:
```py
_T = TypeVar("_T")
_MaybeSequence = _T | Sequence[_T]
_MaybeTypeSequence = Type[_T] | Sequence[Type[_T]]
```

**Reply by erictraut** [reply]
I'd like to investigate this one closer because I think it should theoretically work. 

As I mentioned, mypy also emits a similar error, but that looks like it might be a bug. It does work if you add an explicit `typing.TypeAlias` annotation, as in `_MaybeSequence: TypeAlias = _T | Sequence[_T]`.

**Reply by erictraut** [reply]
I figured out the root cause of the problem, and it will be addressed in the next release of pyright.

**Reply by povilasb** [reply]
Sweet, thank you, Eric!

`TypeAlias` has no effect for me though :/

**Comment by povilasb** [comment]
Ah, seems like this issue is not present in the `1.1.245` anymore. Thank you, Eric 🙌
